### ISSUE TITLE
Instantiate new "forum" for discussion and pre-baking of issues 

### ISSUE THREAD

**Issue by twMat** [Jermolene/TiddlyWiki5#2262]
Forgive [double posting](https://groups.google.com/forum/#!topic/tiddlywikidev/rHwIltkWBSM) but I feel some urgency on the matter since the (justified!) ambition to clean up among the _Issues_ also would/will stifle an often constructive and needed discussion of ideas;

---

Referring to Jeremys declaration of intention on yesterdays hangout 96, that GitHub issues should be more focussed and actionable and that he's intending to close down issues that don't fulfill this...

...and referring to earlier discussions (a year or two ago) on using GitHub instead of the Google groups...

might this be a time to start a Github "place" that might work as a more formative forum to cook ideas before they make it as issues or PRs as well as for general dev discussion?

IMO there is often value also in non-actionalbe issues and it is not uncommon to see how an issue simply needed some bouncing back and forth before it became actionable. It would be a  mistake to simply close the many not-immediately-actionable issues we have and I would strongly prefer to see them moved instead. Or perhaps start a new place where the actionable ones are [to be].

The Github interface is also superior to the Google groups in terms of editing, searching and @notifyiing.

The google tw-dev group is not very active so maybe bringing it closer to the actual development will be good for that development.

Whaddyasay?

Does github allow for more than one forum in one account? What is possible or not possilbe?

<:-)

---

And, in some ironic sense, I guess I'd better make this post actionable:

Please open up a forum for discussion, ideally that issue-threads can be ported to if they deviate from the requested actionability and conciseness for actual _Issues_.

**Comment by Jermolene** [comment]
Hi @twMat 

> ...and referring to earlier discussions (a year or two ago) on using GitHub instead of the Google groups...

I think that that experiment failed; we've ended up with over 500 open issues that makes them very hard to use for practical purposes.

For the moment we'll revert to using the TWDev mailing list for open ended dev discussions. We can also of course continue to discuss tickets that are closed; being closed isn't intended to stop the discussion, just to signal that the issue cannot actioned at present.

Hopefully by the end of the year we'll be doing all our discussions in federation...

**Comment by twMat** [comment]
Ok, if it wasn't for your final comment that

>Hopefully by the end of the year we'll be doing all our discussions in federation...

...then I'd argue that my OP is exactly about solving the 'failed aspect' of the experiment. I don't think many people have understood that you have the dev discussions in mind as one application area for federation, at least I wasn't sure you had this in mind, so it is valuable that you're stating it explicitly. It should be an incentive for all of us to work on federation.

Closing.

**Comment by tobibeer** [comment]
I wanted to point out that there are different stages to an issue being actionable. Actionable does **not** (necessarily) mean we can start writing code right away. To me it means: It is clear what the next step is and the next step can be undertaken.

Perhaps this is a wee philosophical, but I really like a **why**, **what**, **how** aproach (in that order) whereas the process of "finding answers to" each of these questions renders the bracketing issue actionable:

* **why**
    * the defined goal(s) of the issue
    * all too often those are understood as intrinsic, rather than being clearly formulated or cast into a set of "requirements"
* **what**
    * what is are the elements / functions we plan to implement (as derived of those **why**s)
    * think mockups, even uml'ish diagrams, call it a "spec" ...something that gives a draft of where we're headed before it's turned code
    * some issues may stop here in that the answer is outsourced to *this is plugin territory*, which should be a tag to the issue, imho... so we're clear on what issues are open for / desirable as plugin development
* **how**
    * the actual approach for implementation, could be js code, templates, content, css
    * all in all, the actual "writing code" part of it all yielding readable text-bits that live up to the defined **how**s, **what**s and the **why**s

...and once we arrived at that **how**, we're actually in a **verification** phase whereas some **tests**:

* **verify** that we built the thing right ...starting at the end: **how** <= **what** <= **why**
* **validate** that we built the right thing ...starting at the beginning: **why** => **what** => **how**

I'm thinking the current testing phase is mildly too much of a "if there are problem's, they'll crop up sooner or later"... rather than a more proactive and well defined: This is how we test issue xyz.

If I understood Jeremy right, he'd like to focus more on the **how** of it all and have GitHub be less centered around the **what**s and **why**s... which leaves the question: If we'll be taking this out of the conversation here: How are we going to make sure things don't just end up in a threaded nirvana elsewhere?

**Comment by twMat** [comment]
Actually, the main question is; Is there any other alternative that would be more fruitful. @Jermolene seems to mean we'll be better off if we use the google dev forum. Are you, @tobibeer ,saying this is not the case?

As a non-dev I can only say Github is superior for discussion... but if the developers say it causes problems for actual development...

I found this interesting [github API doc](https://developer.github.com/v3/issues/) where it seems the role of "Issues" has changed... and into more how we use them, if I understand... but on the other hand the [help docs](https://help.github.com/articles/about-issues/) dont' seem to reflect the way we're currently using it. Maybe we're just an unusually active bunch?

>How are we going to make sure things don't just end up in a threaded limbo elsewhere?

Well, in a way you could say the same thing about issues here, no?

As stated in my OP, it would be really good if GitHub allows/enables a "parallel" Issue forum, so one for actionables and one for discussion and with tools to move a thread from one to the other...

...BUT I think TWederation will open up for a superior system for discussion and pre-dev. For one thing, it would let us use features Github lack, not least "TW-specific" features and possibilities to really customize it for our purposes.

...and, what's really exciting is that we're on the very verge of being able to do many of these things already. You can *now right away* include the TWederation plugin @inmysocks and I are working on and subscribe to various peoples TW's so that anything tagged, say, `@issue ` is fetched into your TW... where you can slice, dice, transclude and categorize.

**Comment by tobibeer** [comment]
> As stated in my OP, it would be really good if GitHub allows/enables a "parallel" Issue forum, so one for actionables and one for discussion and with tools to move a thread from one to the other...

I would prefer a tagging approach where we clearly indicate what state an issue is in... very much so in terms of defining the **why**, **what** or **how** of doing it, in other words, there should be no talking **what**s without clearly defined **why**s. Likewise there should be no talking **how**s without clearly defined **what**s. And only if we have the **how**s drafted, should there be actual code committed. Of course, defining the how does not mean to draft the entire code in advance... but it describes the approach and the involved bits of the puzzle, so that architectural problems can be spotted before code is written.

How exhaustive all this is very much depends on the issue. For some issues that may boil down to three lines.

I can see how this may seem far out there. However, I'd think it would be rather straight forward and manageable and not that tough a job at all. We could clearly filter the issue lists to look for issues we're comfortable to work on, e.g. if coding is not your thing, you would not be picking those where a **how** needs defining. So, you'd be looking for issues where drafting **what** to do is the next action to be done.

The OP of each issue would and could (even all-at-once) define these three aspects or stages ...and depending on what tags an issue has — literally "what", "how" or "why" — it would be clear if the next things to define was either the **why**s, the **what**s, the **how**s and only then some eventually committed code-bits via related pull requests.

As for tests, I'd prefer whomever works an issue works on defining those as soon as we get to enumerate the **what**s (here possibly "acceptance criteria")... and any **how**s would be accompanied by defined tests of a perhaps more automated nature.

**Comment by tobibeer** [comment]
So, I would think rather than try and close 500 issues — which is sure one way to go about it — I think it was a better approach to **thoroughly assess** and re-evaluate the state of each issue.

For example, I think issues should have a lot more tags reflecting the "aspect" they're concerned with, e.g. what we're talking about, like "filter", "template", "startup", "localization", "control-panel" "plugin-library", "ui", etc... whatever categories are being used. I would suggest using the GitHub wiki to actually list all those categories and their meaning.

Also, while perhaps of a mildly more dynamic nature, I'd suggest assigning "priority" tags, so as to indicate a rough guesstimate as to how important an issue is, overall, i.e. how useful or critical it is.

Another one could be "complexity", possibly indicating how much work is actually expected to be involved in all that, e.g. anything from 1 hour to even 1 month or so.

**Comment by felixhayashi** [comment]
Hi Tobias,

all very reasonable and systematic. I am also in favor of a tagging and
filtering approach instead of closing issues (which makes them
invisible) just because an issue was lacking a why/what/how which could
be added later on.

So for an issue to be actionable it must contain why && what && how. –
However, the "how" should only be a design sketch, not a PR.

Jeremy could then decide for himself at which he sees a PR by using tags.

-Felix

**Comment by tobibeer** [comment]
All the above involves slightly more work upfront, but I think it significantly improves "actionability", i.e. the transparency as to 'what needs doing'... of course, not doing anything much being another option, surely.

What is important to me is that we find a way to prioritize things and thus also make it better discoverable what to do next.

I think it would help if there was a trello corresponding to the backlog... where all issues, especially new ones get immediately sorted into (in terms of a basic, gut-feeling based importance). Then, there could be a column for each new release, a bit like a rudimentary sprint backlog, so we know:

1. what's in focus right now
2. what' next in the overall backlog

Each release column would possibly start with a card that summarizes the focus of the release.

Rather than trying to reference all issues directly in trello, we could have some sort of session that has us grab issues and reference them in such a 'next release' list... while perhaps immediately listing any suggestions that are not going to be in the next release but otherwise put i the overall backlog.

Since GitHub doesn't provide any backlog / kanban board thingy... I think that is what mostly constrains our abilities of prioritization / communicating priorities.

Any issues thus tracked in trello could be given a backlog tag on GitHub so as to indicate that they're being preferred in terms of actionability.

All this does not mean that any new stuff won't be considered. It just means there's a more transparent idea what to do with it with respect to what's already in focus.

Possibly, those issues with the greatest potential possibly should be higher up the list than others.

Not sure how easy this works or if trello has free plans for opensource projects, but we could make it so that editing that trello board could be more of a community thing than those github issues. Or perhaps the other way around, so that the actual backlog and prioritization is in the sole hands of Jeremy and tagging things on Github becomes a more community driven effort. Not sure which granularity either provides I terms of access rights. All these details could be drafted in some contribute.md.

**Comment by twMat** [comment]
[Reopening]

@tobibeer 

But who would apply these labels? Because regular users can't do this, right? 

>The OP of each issue would and could (even all-at-once) define these three aspects or stages [...]

Can github be made so that clicking "New Issue" opens a *template* that the user fills in? Otherwise I'm skeptical to demand users to author their posts following a certain format - github is already scary and difficullt to people.

>trello

...yet another system? :-/

**Comment by tobibeer** [comment]
> Can GitHub be made so that clicking "New Issue" opens a template that the user fills in?

That would be overkill. I'd rather think of it the other way around. Issues are created and those taking it upon themselves to moderate issues do the tagging ...and possibly cross-referencing in some trello backlog.

> I'm skeptical to demand users to author their posts following a certain format - github is already scary and difficullt to people.

Of course, not every user would be expected to do that. To me, this is the barrier that needs to be established so as to not have GitHub be a discussion forum. On the other hand, an issue can always start in a poorly defined manner. What I find missing the most atm is an ability for a "team description" that is bracketing an issue.

So, some user may create an issue, but the team gets to cast that into a bracketing / well defined form / spec, so that it becomes a lot more actionable and clear as to what the status is. One way to achieve this would be to, in fact, open a "team issue" for each issue where things are clearly laid out rather than "discussed". These would be the ones being actioned and the other ones be their source(s). If you will, issues that possibly bracket multiple other issues. Another way would be to draft that spec in the GitHub wiki, one page for each issue. So, that would hold the destillation of **why** to do **what** and **how** ...we eventually **test** it all. In other words, these details would actually not be part of the issue thread at all... only the (status) tags would be.

I think Jeremy can afford certain "fictional" roles only some contributors are granted and trusted to adhere to. So, there may be extended privileges granted to some users, but that still does not mean they are to perform all of the things they theoretically could do, like merge PRs etc. It would only mean that as of your role in the team you can...

* assign tags
* possibly edit the content (perhaps in the GitHub wiki itself) that documents those tags / the workflow, if needed
* edit wiki pages that jot the spec for an issue

**Comment by twMat** [comment]
>Another way would be to draft that "spec" in the GitHub wiki, one page for each issue.

I think a wiki type top article with a discussion thread immediately below it would be superior. The article - intended as a spec draft or a summary, could be edited by, ideally, the OP author and others who have proven credible. Ideally some kind of versioning also, saving previous drafts.

For the Github wiki here, I just put up a [test page](https://github.com/Jermolene/TiddlyWiki5/wiki/Test-page-for-Collaborative-Spec-drafting-ref-%232262-%5Blink%5D(https:--github.com-Jermolene-TiddlyWiki5-issues-2262)

(...hupp! that link did not work as markup, good to know. Well, it does work to click on.)

My first impression is that the Wiki application here is not polished enough or full featured for our needs. Other than the explicit tests I do in the article, most which fail, it seems the front interface is insufficient with a clumsy ToC that would get very messy. It is probably more for FAQs or similar.

So, unless we implement a solution using TWederation (which I really think we can and which would be designed according to our exact needs) then the question is if we could use regular Issue threads for collaborative spec drafting.

Returning to the idea mentioned above, ie an editable top post followed by the discussion thread;

1. Can the administrator appoint special rights to certain users....
2. ...so that those special users can *insert* an empty post to be a "thread head"?
3. And can they (or other special users) be given rights to collaboratively edit this post?
4. Perhaps also, if the OP author does structure his proposal well, then the special users could also edit this post, so it becomes a clean cut spec.

(Reflection: In taking a decision to appoint or give special roles to people one must also consider how misuse of role will be treated... ideally without hurting feelings etc.)

**Comment by tobibeer** [comment]
Hi Mat, I'm a little confused by the link you created up there, I've renamed it to simply:

https://github.com/Jermolene/TiddlyWiki5/wiki/2262

**Comment by tobibeer** [comment]
> Can the administrator appoint special rights to certain users...

There's no need.

> ...so that those special users can insert an empty post to be a "thread head" (that has a stricter specification format)?

For that either

> And can they (or other special users) be given rights to collaboratively edit this post?

We simply edit a page named as the issue.

> Perhaps also, if the OP author does structure his proposal well, then the special users could also edit this post, so it becomes a clean cut spec.

No need to edit issues, simply edit the corresponding wiki page.

**Comment by twMat** [comment]
Tobias,

>Hi Mat, I'm a little confused by the link you created up there, I've renamed it to simply:

Hm, I'm thinking you or someone might have deleted the page I created or did you read the text with the 2-3 tests? The apparence of the link I posted above did look "strange" because I attempted to include a link in the title of the wikipage, to see if it would work (it didn't).

---

OK, so for your proposal to be used, @Jermolene and users would have to adopt it, i.e that;

- issues are in effect accepted to be discussion threads
- an issue-specific wiki page is created by anyone if there is a need (which should be the typical case) 
- ...where an actionable spec is gradually refined by anyone/everyone
- and this wiki page is (probably) the place to go when implementing something for a PR

@Jermolene - good/bad/worth-a-try-if...?


The wiki page would likely also swerve a the best starting point when jumping into a on issue thread.

I suspect there will be multiple variants for specs on an issue when opinions diverge, which should be OK, i.e 2262, 2262b, etc.

Because the wiki ToC lists alphabetically, we might even be able to use some kind of prio prefix or grouping if issues.

Potential problem: How to protect against sabotage, intentional or not? Someone dislikes a page or dislikes a detail in a spec and changes it without calling attention to it, instead of creating a new one... ?
